File: mcp-server/src/keychain.rs
use crate::error::{Result, ValtError};

/// Unified keychain service name shared between CLI and MCP server.
const SERVICE: &str = "valt";

/// Legacy MCP service name kept for backward-compat migration.
const LEGACY_SERVICE: &str = "valt-mcp-server";

const USER: &str = "auth-token";

const AGENT_TOKEN_USER: &str = "valt-agent";
const AGENT_TOKEN_KEY: &str = "valt-agent-token";

/// get_token returns the stored auth token, checking env var first,
/// then the unified "valt" keychain entry, then the legacy "valt-mcp-server" entry.
pub fn get_token() -> Option<String> {
    // Check env var first (useful for CI/headless).
    if let Ok(token) = std::env::var("VALT_AUTH_TOKEN") {
        if !token.is_empty() {
            return Some(token);
        }
    }
    // Try unified service name (written by CLI after valt setup).
    if let Some(token) = keyring::Entry::new(SERVICE, USER)
        .ok()
        .and_then(|e| e.get_password().ok())
        .filter(|t| !t.is_empty())
    {
        return Some(token);
    }
    // Fallback: try legacy MCP service name for backward compatibility.
    keyring::Entry::new(LEGACY_SERVICE, USER)
        .ok()
        .and_then(|e| e.get_password().ok())
        .filter(|t| !t.is_empty())
}

#[allow(dead_code)]
pub fn store_token(token: &str) -> Result<()> {
    let entry = keyring::Entry::new(SERVICE, USER)
        .map_err(|e| ValtError::Keychain(e.to_string()))?;
    entry.set_password(token)
        .map_err(|e| ValtError::Keychain(e.to_string()))
}

#[allow(dead_code)]
pub fn delete_token() -> Result<()> {
    // Delete from unified service name (ignore not-found).
    let _ = keyring::Entry::new(SERVICE, USER)
        .ok()
        .and_then(|e| e.delete_credential().ok());
    // Delete from legacy service name (ignore not-found).
    let _ = keyring::Entry::new(LEGACY_SERVICE, USER)
        .ok()
        .and_then(|e| e.delete_credential().ok());
    Ok(())
}

pub fn get_agent_token() -> Result<Option<String>> {
    // Check env var first (useful for CI/headless).
    if let Ok(token) = std::env::var("VALT_AGENT_TOKEN") {
        if !token.is_empty() {
            return Ok(Some(token));
        }
    }
    // Try OS keychain; not-found is not an error.
    let token = keyring::Entry::new(AGENT_TOKEN_USER, AGENT_TOKEN_KEY)
        .ok()
        .and_then(|e| e.get_password().ok())
        .filter(|t| !t.is_empty());
    Ok(token)
}

pub fn set_agent_token(token: &str) -> Result<()> {
    let entry = keyring::Entry::new(AGENT_TOKEN_USER, AGENT_TOKEN_KEY)
        .map_err(|e| ValtError::Keychain(e.to_string()))?;
    entry.set_password(token)
        .map_err(|e| ValtError::Keychain(e.to_string()))
}

#[allow(dead_code)]
pub fn delete_agent_token() -> Result<()> {
    let entry = keyring::Entry::new(AGENT_TOKEN_USER, AGENT_TOKEN_KEY)
        .map_err(|e| ValtError::Keychain(e.to_string()))?;
    entry.delete_credential()
        .map_err(|e| ValtError::Keychain(e.to_string()))
}

File: valt-cli/internal/keychain/keychain.go
package keychain

import (
	"fmt"
	"os"

	"github.com/zalando/go-keyring"
)

// service is the unified keychain service name shared between CLI and MCP server.
const service = "valt"

// legacyService is the old CLI service name kept for backward-compat migration.
const legacyService = "valt-cli"

const userKey = "auth-token"

// GetToken returns the stored auth token, checking env var first,
// then the unified "valt" keychain entry, then the legacy "valt-cli" entry.
func GetToken() (string, error) {
	if v := os.Getenv("VALT_AUTH_TOKEN"); v != "" {
		return v, nil
	}
	// Try new unified service name first.
	token, err := keyring.Get(service, userKey)
	if err == nil && token != "" {
		return token, nil
	}
	// Fallback: try legacy service name for backward compatibility.
	token, err = keyring.Get(legacyService, userKey)
	if err == nil && token != "" {
		// Migrate: re-store under new service name.
		_ = keyring.Set(service, userKey, token)
		return token, nil
	}
	return "", nil // not found = not logged in
}

// SetToken stores the auth token in the OS keychain under the unified service name.
func SetToken(token string) error {
	if err := keyring.Set(service, userKey, token); err != nil {
		return fmt.Errorf("storing token in keychain: %w", err)
	}
	return nil
}

// DeleteToken removes the stored token from both the unified and legacy service names.
func DeleteToken() error {
	// Delete from new service name (ignore not-found errors).
	_ = keyring.Delete(service, userKey)
	// Delete from legacy service name (ignore not-found errors).
	_ = keyring.Delete(legacyService, userKey)
	return nil
}
